src/decays/decay_amplitudes.cpp

```diff
@@ -0,0 +1,336 @@
+   Decay_amplitude_SSS amp;
+   amp.m_decay = amp2.m_decay;
+   amp.m_scalar_1 = amp2.m_scalar_1;
+   amp.m_scalar_2 = amp2.m_scalar_2;
+   amp.form_factor = factor * amp2.form_factor;
+   return amp;
+}
+Decay_amplitude_SSS operator*(Decay_amplitude_SSS const& amp2, std::complex<double> factor) {
+   return operator*(factor, amp2);
+}
+
+Decay_amplitude_SSV operator*(std::complex<double> factor, Decay_amplitude_SSV const& amp2) {
+   Decay_amplitude_SSV amp;
+   amp.m_decay = amp2.m_decay;
+   amp.m_scalar = amp2.m_scalar;
+   amp.m_vector = amp2.m_vector;
+   amp.form_factor = factor * amp2.form_factor;
+   return amp;
+}
+Decay_amplitude_SSV operator*(Decay_amplitude_SSV const& amp2, std::complex<double> factor) {
+   return operator*(factor, amp2);
+}
+
+Decay_amplitude_SVV operator* (std::complex<double> factor, Decay_amplitude_SVV const& amp2) {
+   Decay_amplitude_SVV amp;
+   amp.m_decay = amp2.m_decay;
+   amp.m_vector_1 = amp2.m_vector_1;
+   amp.m_vector_2 = amp2.m_vector_2;
+   amp.massless_vector_threshold = amp2.massless_vector_threshold;
+   amp.form_factor_g = factor * amp2.form_factor_g;
+   amp.form_factor_11 = factor * amp2.form_factor_11;
+   amp.form_factor_12 = factor * amp2.form_factor_12;
+   amp.form_factor_21 = factor * amp2.form_factor_21;
+   amp.form_factor_22 = factor * amp2.form_factor_22;
+   amp.form_factor_eps = factor * amp2.form_factor_eps;
+   return amp;
+}
+Decay_amplitude_SVV operator*(Decay_amplitude_SVV const& amp2, std::complex<double> factors) {
+   return operator*(factors, amp2);
+}
+
+Decay_amplitude_SFF operator*(std::complex<double> factor, Decay_amplitude_SFF const& amp2) {
+   Decay_amplitude_SFF amp;
+   amp.m_decay = amp2.m_decay;
+   amp.m_fermion_1 = amp2.m_fermion_1;
+   amp.m_fermion_2 = amp2.m_fermion_2;
+   amp.form_factor_left = factor * amp2.form_factor_left;
+   amp.form_factor_right = factor * amp2.form_factor_right;
+   return amp;
+}
+Decay_amplitude_SFF operator*(Decay_amplitude_SFF const& amp, std::complex<double> factor) {
+   return operator*(factor, amp);
+}
+
+double Decay_amplitude_SFF::square() const
+{
+   const double m_in_sq = Sqr(m_decay);
+   const double m_1_sq = Sqr(m_fermion_1);
+   const double m_2_sq = Sqr(m_fermion_2);
+
+   return (m_in_sq - m_1_sq - m_2_sq) *
+      (AbsSqr(form_factor_left) + AbsSqr(form_factor_right))
+      - 4. * m_fermion_1 * m_fermion_2 * Re(
+         form_factor_left * Conj(form_factor_right));
+}
+
+double Decay_amplitude_FFS::square() const
+{
+   const double m_in_sq = Sqr(m_decay);
+   const double m_f_sq = Sqr(m_fermion);
+   const double m_s_sq = Sqr(m_scalar);
+
+   return 0.5 * (m_in_sq + m_f_sq - m_s_sq) *
+      (AbsSqr(form_factor_left) + AbsSqr(form_factor_right))
+      + 2. * m_fermion * m_scalar * Re(
+         form_factor_left * Conj(form_factor_right));
+}
+
+// @todo handle massless vectors safely
+// @todo check these expressions, they appear not to agree with SARAH/SPheno
+double Decay_amplitude_FFV::square() const
```

Comment:
This todo comes from Dylan. I'm not sure what he was concerned about.

Reply:
From memory when comparing the expressions to the implementation in SARAH/SPheno they did not appear to be equivalent, and so I did not trust them. Given that these are for fermion decays (and so aren't supported/haven't been tested numerically?), they could be removed until the decays code is extended to support fermion and vector decays?

Reply:
Just to clarify, the suggestion would be to remove the unused fermion decay routines (i.e., the structs `Decay_amplitude_FFV` and `Decay_amplitude_FFS` and the associated functions), not just the comments. It is easy enough to revert in a separate branch when fermion and vector decays are implemented down the line, and avoids introducing untested code into the stable release. The other option would be to properly test them, but maybe this can be done later given that this release is only supporting scalar decays?

Reply:
I think we could either keep the code but adapt the comment to make it clear the code is not currently used and is WIP for a future version or we could follow Dylan’s suggestion and remove the routines until they are ready.   I’ll try to suggest something for the corner tonight, after I get home.

Reply:
OK pushed extra comments above to explain, but i now see similar todo statements above Decay_amplitude_SVV, 

```
// @todo handle massless vectors safely
// @todo check expressions correct
double Decay_amplitude_SVV::square() const
```

but this routine should be used right?  Have we checked its correct?

Reply:
The last part of this routine, concerning decays of scalar in massive, on-shell vectors have not been thoroughly checked beyond numerical comparison in heavy higgs decays.

Reply:
OK so the issue is there was  S->VV when we looked squark decays so we only tested it for Higgs, not a generic scalar?  But there is no reason to doubt it would be correct right? So maybe we could explicitly state that in the comment, ie replace the comment with  

"@todo numerically validate decays of generic scalar to two massive on-shell decays.  So far we have not studied an example like this."  ?

Reply:
Also sorry if I am missing something obvious, but what does the other comment mean here about handling massless vectors safely? we have decays into gamma gamma and into gamma Z, so thus is presumably done ?  or there is a special case issue?

Reply:
So doesn't that mean this todo is already addressed and can be removed?  I mean checking that its smaller than `std::numeric_limits<double>::epsilon()` is the right thing to do, no?

What about my other comment do you agree with it? can I change that todo like this?

Reply:
All other comments were marked as resolved. I think they're addressed.

Reply:
I mean the comment right above the one you replied to, ie this:



OK so the issue is there was no S->VV when we looked squark decays, so we only tested it for heavy Higgs, not a generic scalar? But there is no reason to doubt it would be correct right? So maybe we could explicitly state that in the comment, ie replace the comment with

"@todo numerically validate decays of generic scalar to two massive on-shell decays. So far we have not studied an example like this." ?

Reply:
One might have a model were photon is part of a multiplet and it's 0 mass comes as an eigenvalue of a mass matrix. Then this check might fail, case it's mass might be quite different from 0.

Reply:
You mean the numerical errors in the diagonalsiation can lead to a non-zero photon mass above the limit, when we know it must actually be zero and should be treated as massless?

Reply:
How about:
`@todo In future versions handle tricky corner cases where e.g. the photon mass is obtained from diagonalisation of a mass matrix and can numerically be non-zero due to diagonalisiaton precision.  Currently this would be treated as a massive vector. `
Its a bit long so i'll try to shorten it if I can can, but the point is readers should see the todos don't mean we released unfinished code and the comment should make clear to all developers what needs doing (I couldn't tell what actually needed doing from the current version).

Reply:
Ok.

Reply:
OK I made a change like that.  i also implemented my suggestion for the other todo, though you did not reply on that. I will approve the pull request now, but you may want to check this before you do the merge.